src/broker/cas_cgw.c

```diff
@@ -2438,3 +2456,260 @@ cgw_utype_to_string (int type)
+
+  SQL_CHK_ERR (hstmt, SQL_HANDLE_STMT, err_code = SQLPrepare (hstmt, (SQLCHAR *) inline_view, SQL_NTS));
+
+  err_code = cgw_get_num_cols (hstmt, &odbc_num_cols);
+  if (err_code < 0)
+    {
+      err_code = ER_FAILED;
+      goto ODBC_ERROR;
+    }
+
+  if (odbc_num_cols == 0)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_PT_SEMANTIC, 2, "Invalid column.", "");
+      err_code = ER_FAILED;
+      goto ODBC_ERROR;
+    }
+
+  if (src_num_cols != odbc_num_cols)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_CGW_INVALID_COLUMNS_SPECIFIED, 0);
+      err_code = ER_FAILED;
+      goto ODBC_ERROR;
+    }
+
+  new_inline_view_len += REWRITE_SELECT_LEN;
+  new_inline_view_len += (int) (strlen (" AS ") * odbc_num_cols);
+  new_inline_view_len += inline_view_len;
+  new_inline_view_len += cols_type_len;
+  new_inline_view_len += odbc_num_cols;
+  new_inline_view_len += REWRITE_FROM_LEN;
+
+  for (int col_num = 1; col_num <= odbc_num_cols; col_num++)
+    {
+      SQL_CHK_ERR (hstmt,
+		   SQL_HANDLE_STMT,
+		   err_code = SQLColAttribute (hstmt,
+					       col_num,
+					       SQL_DESC_NAME, col_name, sizeof (col_name), &col_name_length, NULL));
+
+      new_inline_view_len += col_name_length;
+    }
+
+  new_inline_view = (char *) MALLOC (new_inline_view_len + 1);
+  strcpy (new_inline_view, "(SELECT ");
+
+  for (int col_num = 1; col_num <= odbc_num_cols; col_num++)
+    {
+      SQL_CHK_ERR (hstmt,
+		   SQL_HANDLE_STMT,
+		   err_code = SQLColAttribute (hstmt,
+					       col_num,
+					       SQL_DESC_NAME, col_name, sizeof (col_name), &col_name_length, NULL));
+
+      if (cols_list[col_num - 1] == NULL || strcmp (cols_list[col_num - 1], "") == 0)
+	{
+	  err_code = REWRITE_FAILED;
+	  goto ODBC_ERROR;
+	}
+
+      strcat (new_inline_view, col_name);
+      strcat (new_inline_view, " AS ");
+      strcat (new_inline_view, cols_list[col_num - 1]);
+
+      if (col_num <= odbc_num_cols - 1)
+	{
+	  strcat (new_inline_view, ", ");
+	}
+      else
+	{
+	  strcat (new_inline_view, " FROM ");
+	  strcat (new_inline_view, inline_view);
+	  strcat (new_inline_view, ")");
+	}
+    }
+
+  rewrite_query = (char *) MALLOC (REWRITE_SELECT_FROM_LEN + new_inline_view_len + where_len + 1);
+  memset (rewrite_query, 0x0, REWRITE_SELECT_FROM_LEN + new_inline_view_len + where_len + 1);
+
+  strncpy (rewrite_query, select_from, select_from_len);
+  strcat (rewrite_query, " ");
+  strcat (rewrite_query, new_inline_view);
+  strcat (rewrite_query, " ");
+  strcat (rewrite_query, where);
+
+  *rewrite_sql = rewrite_query;
+
+  if (hstmt)
+    {
+      SQLFreeHandle (SQL_HANDLE_STMT, hstmt);
+    }
+
+  FREE_MEM (source);
+
+  if (cols_list)
+    {
+      cgw_free_string_array (cols_list);
+    }
+
+  return err_code;
+
+ODBC_ERROR:
+  if (hstmt)
+    {
+      SQLFreeHandle (SQL_HANDLE_STMT, hstmt);
+    }
+
+  FREE_MEM (source);
+
+  if (cols_list)
+    {
+      cgw_free_string_array (cols_list);
+    }
+
+
+  FREE_MEM (rewrite_query);
```

Comment:
Why don't you free **new_inline_view** ?

Reply:
Thank you for comment.